### ISSUE TITLE
[BUG] azure-messaging-eventhubs 5.3.1 fails with proxy 

### ISSUE THREAD

**Issue by jimwrightcz** [Azure/azure-sdk-for-java#18334]
**Describe the bug**
We are able to publish through azure-messaging-eventhubs 5.2.0 but not 5.3.1.

***Exception or Stack Trace***

Note:
* I have replaced our namespace and eventhub name with ....
* To reproduce on my workstation I am using a command like ``$ ssh -L 3128:proxyHost:3128 appServer`` but we get the same behaviour using a similar test harness on app servers
* This is with 5.3.1 and with 5.2.0 there are no warnings such as "No subject alternative DNS name matching localhost found."

````
[16:19:00.867,INFO ,amqp.ProxyOptions,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  Username or password is null. Using system-wide authentication.
[16:19:01.044,INFO ,eventhubs.EventHubClientBuilder,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  connectionId[MF_37692f_1608736741042]: Emitting a single connection.
[16:19:01.075,INFO ,implementation.EventHubConnectionProcessor,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  namespace[....servicebus.windows.net] entityPath[...]: Setting next AMQP channel.
[16:19:01.075,INFO ,implementation.EventHubConnectionProcessor,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  namespace[....servicebus.windows.net] entityPath[...]: Next AMQP channel received, updating 0 current subscribers
[16:19:01.101,INFO ,implementation.ReactorConnection,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  connectionId[MF_37692f_1608736741042]: Creating and starting connection to localhost:3128
[16:19:01.124,INFO ,implementation.ReactorExecutor,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  connectionId[MF_37692f_1608736741042], message[Starting reactor.]
[16:19:01.134,INFO ,handler.WebSocketsProxyConnectionHandler,td:single-1,dc:]
  onConnectionInit host[....servicebus.windows.net:443], connectionId[MF_37692f_1608736741042]
[16:19:01.135,INFO ,handler.ReactorHandler,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042] reactor.onReactorInit
[16:19:01.136,INFO ,handler.ConnectionHandler,td:single-1,dc:]
  onConnectionLocalOpen hostname[....servicebus.windows.net:443], connectionId[MF_37692f_1608736741042], errorCondition[null], errorDescription[null]
[16:19:01.144,INFO ,implementation.AzureTokenManagerProvider,td:jimwrightcz.PublishEventsWithWebSocketsAndProxy.main(),dc:]
  Creating new token manager for audience[amqp://....servicebus.windows.net/...], resource[...]
[16:19:01.202,INFO ,handler.ConnectionHandler,td:single-1,dc:]
  onConnectionBound hostname[localhost], connectionId[MF_37692f_1608736741042]
[16:19:01.367,INFO ,proxy.ProxyConfiguration,td:single-1,dc:]
  username or password is null. Using system-wide authentication.
[16:19:01.370,INFO ,handler.WebSocketsProxyConnectionHandler,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042] addProxyHandshake: hostname[....servicebus.windows.net:443]
[16:19:01.378,INFO ,impl.ProxyImpl,td:single-1,dc:]
  Writing proxy request:
CONNECT ....servicebus.windows.net:443 HTTP/1.1
Host: ....servicebus.windows.net:443
Connection: Keep-Alive
Dec 23, 2020 4:19:01 PM org.apache.qpid.proton.engine.impl.ssl.SimpleSslTransportWrapper process
WARNING: No subject alternative DNS name matching localhost found.
[16:19:01.769,WARN ,handler.ConnectionHandler,td:single-1,dc:]
  onTransportError hostname[....servicebus.windows.net:443], connectionId[MF_37692f_1608736741042], error[javax.net.ssl.SSLHandshakeException: No subject alternative DNS name matching localhost found.]
[16:19:01.773,INFO ,implementation.EventHubReactorAmqpConnection,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042]: Disposing of connection.
[16:19:01.773,INFO ,implementation.ReactorConnection,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042], errorCondition[n/a]: Disposing of ReactorConnection.
[16:19:01.773,INFO ,implementation.ReactorConnection,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042]: Removing session '...'
[16:19:01.773,INFO ,implementation.ReactorSession,td:single-1,dc:]
  connectionId[MF_37692f_1608736741042], sessionId[...], errorCondition[n/a]: Disposing of session.
[16:19:01.775,INFO ,implementation.AmqpExceptionHandler,td:single-1,dc:]
  Shutdown received: ReactorExecutor.close() was called., isTransient[false], initiatedByClient[true]
[16:20:01.171,INFO ,implementation.RetryUtil,td:parallel-1,dc:]
  TimeoutException error occurred. Retrying operation. Attempt: 1.
[WARNING] 
java.lang.IllegalStateException: Timeout on blocking read for 60000 MILLISECONDS
    at reactor.core.publisher.BlockingSingleSubscriber.blockingGet (BlockingSingleSubscriber.java:123)
    at reactor.core.publisher.Mono.block (Mono.java:1704)
    at com.azure.messaging.eventhubs.EventHubProducerClient.createBatch (EventHubProducerClient.java:126)
    at jimwrightcz.PublishEventsWithWebSocketsAndProxy.main (PublishEventsWithWebSocketsAndProxy.java:78)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:834)
````

**To Reproduce**
Access my fork here and check my last commit. It's a hack but aims to be minimal:
https://github.com/jimwrightcz/azure-sdk-for-java/tree/azure-messaging-eventhubs-5.3.1.proxy-bug

It starts with the sample provided and you could reproduce by changing the version here:
https://github.com/jimwrightcz/azure-sdk-for-java/blob/azure-messaging-eventhubs-5.3.1.proxy-bug/sdk/eventhubs/azure-messaging-eventhubs/src/samples/pom.xml#L57

if you want to use maven and that POM.

***Code Snippet***
See https://github.com/jimwrightcz/azure-sdk-for-java/blob/azure-messaging-eventhubs-5.3.1.proxy-bug/sdk/eventhubs/azure-messaging-eventhubs/src/samples/hacked/jimwrightcz/PublishEventsWithWebSocketsAndProxy.java

It is a copy of:
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/PublishEventsWithWebSocketsAndProxy.java
but moved to a different package and with the ability to configure namespace, eventhub,  SAS policy and key on the command line, in the POM or by defining system properties when running maven.

**Expected behavior**
With 5.2.1 the output is as above until the warning and then it continues with:
````
[16:15:15.594,INFO ,handler.ConnectionHandler,td:single-1,dc:]
  onConnectionRemoteOpen hostname[....servicebus.windows.net:443], connectionId[MF_5f3a87_1608736514646], remoteContainer[e439da00877d408bb0fbc2737e9ef3e5_G46]
[16:15:15.595,INFO ,implementation.EventHubConnectionProcessor,td:single-1,dc:]
  namespace[....servicebus.windows.net] entityPath[...]: Channel is now active.
...
````
and events are successfully published.

**Screenshots**
none

**Setup (please complete the following information):**
 - OS: RedHat 7.9/Ubuntu 16.04
 - IDE : IntelliJ (output above) but I also reproduced by packaging our own test harness and executing java from the commandline
 - Version of the Library used: 5.3.1

**Additional context**
Despite the warning (which I do not fully understand) I assume this is a bug and not a feature because it works with 5.2.0 and also because of:
https://github.com/Azure/azure-sdk-for-java/issues/5725

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [/] Bug Description Added
- [/] Repro Steps Added
- [/] Setup information Added

**Comment by in-fke** [comment]
Same here with 5.4.0, something that it's trying to verify the proxies host name for SSL
```
2021-02-05T15:56:07,543 WARN  [reactor-executor-1] ssl.SimpleSslTransportWrapper:375 No subject alternative DNS name matching proxy found.
2021-02-05T15:56:07,544 WARN  [reactor-executor-1] handler.ConnectionHandler:344 onTransportError hostname[iothub-ns-in-sphinx-7486724-7d375b9cc5.servicebus.windows.net:443], connectionId[MF_cad884_1612536966447], error[javax.net.ssl.SSLHandshakeException: No subject alternative DNS name matching proxy found.]
2021-02-05T15:56:07,551 INFO  [reactor-executor-1] implementation.EventHubReactorAmqpConnection:338 connectionId[MF_cad884_1612536966447]: Disposing of connection.
```

**Comment by Dak0ta** [comment]
It appears to be expecting the hostname of the proxy to match the certificate returned by the proxy when it should be using the proxy target hostname as the expected hostname.

Ideally this could be handled in ConnectionHandler so that when #17907 is done the expected hostname can be passed on.

If #17907 was done you could at least write your own hostname verifier to use the original host vs the proxy server host instead of bypassing the validation entirely as is requested in #17689

Hitting this as well in the Proxied AMPQ connection but storage doesn't seem to have the same problem.

**Comment by in-fke** [comment]
And I also wonder why Java's ProxySelector.getDefault() is not being honored, but that may be a different topic.

**Comment by conniey** [comment]
Back in 1.6.0, we added SSL verification (https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core-amqp/CHANGELOG.md#new-features-4). The error you're seeing is because it's not getting an SSL certificate to validate. 

Related #17907. 

> And I also wonder why Java's ProxySelector.getDefault() is not being honored, but that may be a different topic.

We use a proxy when `proxyOptions` are set when building the client.

**Comment by jimwrightcz** [comment]
Thanks for the feedback which I now took time to try and understand. It it is possible this warning is distracting from the main issue:
`WARNING: No subject alternative DNS name matching localhost found.`

I don't recall if I got this warning when running a test harness on an app server instead of locally with `ssh -L 3128:proxyHost:3128 appServer` and if not then the warning is nothing more than a feature. (I'm not planning to try again. Based on a colleague's suggestion I tried to suppress it with `-Dcom.microsoft.azure.servicebus.ssl.verifymode=verifyCertificateOnly' without success.)

My first concern is still with `[WARNING] 
java.lang.IllegalStateException: Timeout on blocking read for 60000 MILLISECONDS`.

**Comment by in-fke** [comment]
> The error you're seeing is because it's not getting an SSL certificate to validate.

![image](https://user-images.githubusercontent.com/73228413/113317799-0a823f80-9310-11eb-9cef-34b96b0ca8d1.png)

![image](https://user-images.githubusercontent.com/73228413/113318188-7f557980-9310-11eb-935d-5b00a260e080.png)

Browser seems happy. Would this imply a temporary problem? 
And what does the proxy have to do with endpoint? For me it looked like the SSL verfiication confuses something, because it states

`No subject alternative DNS name matching proxy found`

The name of our proxy happens to be "proxy", so SSL **verification is confusing host names**? (proxy vs. endpoint)

**Comment by in-fke** [comment]
Any progress on this and my theory that SSL verification is confusing host names (proxy vs. endpoint)?

**Comment by Dak0ta** [comment]
Hi @conniey - wasn't this resolved with Event Hubs 5.6.0 and Checkpoint Store Blob 1.5.1?

**Comment by conniey** [comment]
I thought I had resolved this but I can't find the commit. I'll have to take a look at this again.

**Comment by Dak0ta** [comment]
> I thought I had resolved this but I can't find the commit. I'll have to take a look at this again.

I'm pretty sure it's resolved, but I couldn't find the related commits either - thanks!

**Comment by conniey** [comment]
@Dak0ta 
After some sleuthing, I found the corresponding PR (https://github.com/Azure/azure-sdk-for-java/commit/2ef91914db3799776ef29fa966dd8bd3b19c68b7#diff-c9a97ed2b428c171b6864c53d2a98598a22205ebbfdd443c34b77e346ef68fc1R319) and the behaviour it was supposed to fix. To verify the peer details against the message broker (Service Bus/Event Hubs) when using a proxy.

**Comment by conniey** [comment]
@in-fke 
I tested this with my own proxy and it seems to work. Can you close this if it works?

**Comment by in-fke** [comment]
Sorry I am not too familiar with GIT yet - could someone please enlighten me how I (generally) determine since which version of which component this has been fixed? Probably starting at the above PR?

**Comment by ramya-rao-a** [comment]
@conniey I believe version 7.2.0 and above should have this fix, right?

**Comment by conniey** [comment]
All of our releases after April/May would have this fix; 5.7.1 and 5.10.0 (our latest) would have it.
